server/schedule/region_scatterer.go

```diff
@@ -0,0 +1,177 @@
+}
+
+// RegionScatterer scatters regions.
+type RegionScatterer struct {
+	cluster    Cluster
+	classifier namespace.Classifier
+	filters    []Filter
+	selected   *selectedStores
+}
+
+// NewRegionScatterer creates a region scatterer.
+func NewRegionScatterer(cluster Cluster, classifier namespace.Classifier) *RegionScatterer {
+	filters := []Filter{
+		NewStateFilter(),
+		NewHealthFilter(),
+		NewStorageThresholdFilter(),
+	}
+
+	return &RegionScatterer{
+		cluster:    cluster,
+		classifier: classifier,
+		filters:    filters,
+		selected:   newSelectedStores(),
+	}
+}
+
+// Scatter relocates the region.
+func (r *RegionScatterer) Scatter(region *core.RegionInfo) *Operator {
+	if r.cluster.IsRegionHot(region.GetId()) {
+		return nil
+	}
+
+	if len(region.GetPeers()) != r.cluster.GetMaxReplicas() {
+		return nil
+	}
+
+	return r.scatterRegion(region)
+}
+
+func (r *RegionScatterer) scatterRegion(region *core.RegionInfo) *Operator {
+	steps := make([]OperatorStep, 0, len(region.GetPeers()))
+
+	stores := r.collectAvailableStores(region)
+	for _, peer := range region.GetPeers() {
+		if len(stores) == 0 {
+			// Reset selected stores if we have no available stores.
+			r.selected.reset()
+			stores = r.collectAvailableStores(region)
+		}
+
+		if r.selected.put(peer.GetStoreId()) {
+			delete(stores, peer.GetStoreId())
+			continue
+		}
+		newPeer := r.selectPeerToReplace(stores, region, peer)
+		if newPeer == nil {
+			continue
+		}
+
+		// Remove it from stores and mark it as selected.
+		delete(stores, newPeer.GetStoreId())
+		r.selected.put(newPeer.GetStoreId())
+
+		op := CreateMovePeerOperator("scatter-peer", region, OpAdmin,
+			peer.GetStoreId(), newPeer.GetStoreId(), newPeer.GetId())
+		steps = append(steps, op.steps...)
+	}
+
+	if len(steps) == 0 {
+		return nil
+	}
+	return NewOperator("scatter-region", region.GetId(), OpAdmin, steps...)
+}
+
+func (r *RegionScatterer) selectPeerToReplace(stores map[uint64]*core.StoreInfo, region *core.RegionInfo, oldPeer *metapb.Peer) *metapb.Peer {
+	// scoreGuard guarantees that the distinct score will not decrease.
+	regionStores := r.cluster.GetRegionStores(region)
+	sourceStore := r.cluster.GetStore(oldPeer.GetStoreId())
+	scoreGuard := NewDistinctScoreFilter(r.cluster.GetLocationLabels(), regionStores, sourceStore)
+
+	candidates := make([]*core.StoreInfo, 0, len(stores))
+	for _, store := range stores {
+		if scoreGuard.FilterTarget(r.cluster, store) {
+			continue
+		}
+		candidates = append(candidates, store)
+	}
+
+	if len(candidates) == 0 {
+		return nil
+	}
+
+	target := candidates[rand.Intn(len(candidates))]
+	newPeer, err := r.cluster.AllocPeer(target.GetId())
+	if err != nil {
+		return nil
+	}
+	return newPeer
+}
+
+func (r *RegionScatterer) collectAvailableStores(region *core.RegionInfo) map[uint64]*core.StoreInfo {
+	namespace := r.classifier.GetRegionNamespace(region)
+	filters := []Filter{
+		r.selected.newFilter(),
+		NewExcludedFilter(nil, region.GetStoreIds()),
```

Comment:
Consider we have 6 stores and many regions all on the same 3 stores, after scatter the 100 regions, all regions will be move to other 3 stores.

Reply:
It will not work like that, check https://github.com/pingcap/pd/pull/890/files#diff-1980e052a5d29b89f9345cc000bb9db8R110.
If the store of a peer is available, it will just assign the same store to the peer.
So with 6 stores and many regions all on the same 3 stores, half of the regions will be assigned to the same 3 stores, another half will be assigned to the other 3 stores.
I think I should add a comment there.

Reply:
:smile:  Got it.